File: backend/routers/tasks.py
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.config import settings
from backend.dependencies import get_current_user, get_db
from backend.models.user import User
from backend.schemas.task import TaskCreate, TaskResponse, TaskUpdate
from backend.services import task_service

router = APIRouter()


@router.get("/", response_model=list[TaskResponse])
def list_tasks(
    status: Optional[str] = Query(None),
    task_type: Optional[str] = Query(None),
    priority: Optional[str] = Query(None),
    property_id: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db),
):
    return task_service.list_tasks(db, current_user.id, status, task_type, priority, property_id, search)


@router.post("/", response_model=TaskResponse, status_code=201)
def create_task(
    data: TaskCreate,
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db),
):
    return task_service.create_task(db, current_user.id, data)


@router.get("/{task_id}", response_model=TaskResponse)
def get_task(
    task_id: uuid.UUID,
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db),
):
    return task_service.get_task(db, current_user.id, task_id)


@router.patch("/{task_id}", response_model=TaskResponse)
def update_task(
    task_id: uuid.UUID, data: TaskUpdate,
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db),
):
    return task_service.update_task(db, current_user.id, task_id, data)


@router.delete("/{task_id}", status_code=204)
def delete_task(
    task_id: uuid.UUID,
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db),
):
    task_service.delete_task(db, current_user.id, task_id)

File: lib/api/tasks.ts
import { api } from "./client";

export interface TaskData {
  id: string;
  property_id?: string | null;
  user_id: string;
  title: string;
  description?: string | null;
  task_type: string;
  due_date?: string | null;
  priority: string;
  status: string;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskCreateData {
  title: string;
  description?: string;
  property_id?: string;
  task_type?: string;
  due_date?: string | null;
  priority?: string;
}

export interface TaskUpdateData {
  title?: string;
  description?: string | null;
  property_id?: string | null;
  task_type?: string;
  due_date?: string | null;
  priority?: string;
  status?: string;
}

export function listTasks(params?: { status?: string; task_type?: string; priority?: string; property_id?: string; search?: string }): Promise<TaskData[]> {
  const sp = new URLSearchParams();
  if (params?.status) sp.set("status", params.status);
  if (params?.task_type) sp.set("task_type", params.task_type);
  if (params?.priority) sp.set("priority", params.priority);
  if (params?.property_id) sp.set("property_id", params.property_id);
  if (params?.search) sp.set("search", params.search);
  const qs = sp.toString();
  return api.get(`/tasks/${qs ? `?${qs}` : ""}`);
}

export function createTask(data: TaskCreateData): Promise<TaskData> {
  return api.post("/tasks/", data);
}

export function updateTask(id: string, data: TaskUpdateData): Promise<TaskData> {
  return api.patch(`/tasks/${id}`, data);
}

export function deleteTask(id: string): Promise<void> {
  return api.delete(`/tasks/${id}`);
}
